# question 1 -  

##### A company conducted a survey of its employees and found that 70% of the employees use the company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the probability that an employee is a smoker given that he/she uses the health insurance plan?

given that out of 100 people ,  we have 70 who have insuracne plan.

out of those 70, we have 40 % people who are smokers

this implies that

people who smoke and have insurance = 40/100 * 70 = 28

so that probability that an employee is a smoker given that he/she uses the health insurance plan is 0.28.

# question 2 - diff btw bernoulli and multinomial naive bayes

Bernoulli Naive Bayes and Multinomial Naive Bayes are two variants of the Naive Bayes classifier, and they are primarily used for different types of data and classification tasks. The main difference between them lies in the type of data they are designed to handle and the underlying assumptions they make.

1. **Bernoulli Naive Bayes**:

   - **Binary Data**: Bernoulli Naive Bayes is specifically designed for binary or binary-like data, where each feature can take on one of two values, typically 0 and 1. It's suitable for situations where you want to model the presence or absence of specific features or characteristics.

   - **Example**: It's commonly used in text classification tasks, where you represent documents as binary vectors, indicating the presence (1) or absence (0) of certain words in the document.

   - **Assumption**: It assumes that each feature is binary and that the features are conditionally independent given the class label.

   - **Probability Estimation**: It models the probability of each feature being 1 (presence) or 0 (absence) in each class.

2. **Multinomial Naive Bayes**:

   - **Count Data**: Multinomial Naive Bayes is suitable for data with discrete features where the values represent counts or frequencies. It's often used in text classification tasks where you have counts of word occurrences or other discrete features.

   - **Example**: In text classification, you might represent documents as vectors of word frequencies, indicating how many times each word appears in the document.

   - **Assumption**: It assumes that the features are multinomially distributed (i.e., they represent counts or frequencies) and that the features are conditionally independent given the class label.

   - **Probability Estimation**: It models the probability of observing a particular count or frequency of each feature in each class.

In summary, the choice between Bernoulli Naive Bayes and Multinomial Naive Bayes depends on the nature of your data and how you've represented your features. If your data is binary or binary-like, Bernoulli Naive Bayes may be more appropriate. If your data consists of counts or frequencies, Multinomial Naive Bayes is generally a better choice. Additionally, you should consider the specific requirements of your classification task and experiment with both algorithms to determine which one performs better on your dataset.

# question 3 -  how can we deal with missing values in case of bernoulli naive bayes

Bernoulli Naive Bayes, like other Naive Bayes variants, can handle missing values to some extent, but the way it handles them depends on how the missing values are represented and how you choose to handle them in your data preprocessing.

Here are a few considerations for handling missing values with Bernoulli Naive Bayes:

1. **Representation of Missing Values**:
   - In Bernoulli Naive Bayes, features are typically binary, representing the presence (1) or absence (0) of a particular characteristic or event.
   - You can choose how to represent missing values. Common approaches include:
     - Assigning a specific value (e.g., -1 or NaN) to indicate missingness.
     - Imputing missing values with a default value (e.g., 0 or 0.5) to indicate uncertainty or neutrality.
     - Creating an additional binary feature to explicitly indicate whether a value is missing or not.

2. **Handling Missing Data During Training**:
   - During the training phase, Bernoulli Naive Bayes can work with missing values by treating them as an additional category or class within each feature.
   - For example, if you use a specific value (e.g., -1) to represent missingness, you can consider it as a third category along with 0 and 1 when estimating probabilities.

3. **Handling Missing Data During Prediction**:
   - When making predictions for instances with missing values, Bernoulli Naive Bayes can still provide a classification based on the available information.
   - You can use the available features to calculate probabilities for each class as usual, and the missing features (represented by -1 or similar) are treated as if they were observed with their respective values.
   - However, this approach assumes that the missingness of a feature is not informative by itself. In many cases, you might assume that the feature is missing at random (MAR).

4. **Imputation Strategies**:
   - Depending on your dataset and the nature of missingness, you may also consider imputation strategies to fill in missing values. Imputation techniques can include mean imputation, median imputation, or more advanced methods like k-Nearest Neighbors (k-NN) imputation.

It's important to note that Bernoulli Naive Bayes, like other Naive Bayes classifiers, relies on the assumption of feature independence given the class. Therefore, how you handle missing values can impact the validity of this assumption. If missingness is related to the class label or other features in a non-random way, it can introduce bias into your model. Careful preprocessing and consideration of missing data mechanisms are essential when using Bernoulli Naive Bayes or any other classification algorithm.

# question 4 - gaussian naive bayes

Yes, Gaussian Naive Bayes can be used for multi-class classification tasks. Gaussian Naive Bayes is a variant of the Naive Bayes classifier that is well-suited for data where features are continuous and assumed to follow a Gaussian (normal) distribution. While it's often used for binary classification problems, it can also be extended to handle multi-class classification tasks.

In multi-class classification with Gaussian Naive Bayes, you typically follow one of two common approaches:

1. **One-vs-Rest (OvR) or One-vs-All (OvA)**:
   - In this approach, you create multiple binary classifiers, one for each class. Each classifier is trained to distinguish between one class and the rest of the classes.
   - For example, if you have three classes (A, B, and C), you would create three binary classifiers:
     - Classifier 1: Classify A vs. (B and C)
     - Classifier 2: Classify B vs. (A and C)
     - Classifier 3: Classify C vs. (A and B)
   - During prediction, you apply each classifier to the input data, and the class with the highest probability becomes the predicted class.

2. **Softmax or Multinomial Distribution**:
   - In this approach, you modify the Gaussian Naive Bayes model to directly handle multi-class classification without breaking it down into binary problems.
   - Instead of using binary classifiers, you calculate the conditional probabilities of each class directly using the Gaussian distribution. These probabilities are then normalized using the softmax function to obtain the final class probabilities.
   - The class with the highest probability is chosen as the predicted class.

The choice between these two approaches often depends on the specific problem and the dataset. The OvR approach is straightforward to implement and can work well when you have a small number of classes. The softmax approach, on the other hand, directly models multi-class probabilities and is commonly used for tasks with a larger number of classes.

In summary, Gaussian Naive Bayes can indeed be adapted for multi-class classification, and the choice of approach depends on the problem's requirements and the number of classes you're dealing with.

# question 5 -- Assignment

In [1]:
import pandas as pd
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [3]:
df = pd.read_csv('/Users/damodargupta/Desktop/projects/spam classification using naive bayes/spambase/spambase.data' )
df

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0                0.00            0.00  ...        0.000        0.000   
1                0.00            0.94  ...        0.000        0.132   
2                0.64            0.25  ...        0.010        0.143   
3                0.31            0.63  ...        0.000        0.137   
4                0.31            0.63  ...        0.000        0.135   
...               ...             ...  ...          ...          ...   
4596             0.00            0.00  ...        0.000        0.232   
4597             0.00            0.00  ...        0.000        0.000   
4598             0.00            0.00  ...        0.102        0.718   
4599             0.00            0.00  ...        0.000        0.057   
4600             0.00            0.00  ...        0.000        0.000   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0             0.0        0.778        0.000        0.000   
1             0.0        0.372        0.180        0.048   
2             0.0        0.276        0.184        0.010   
3             0.0        0.137        0.000        0.000   
4             0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...   
4596          0.0        0.000        0.000        0.000   
4597          0.0        0.353        0.000        0.000   
4598          0.0        0.000        0.000        0.000   
4599          0.0        0.000        0.000        0.000   
4600          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537                          40   
...                          ...                         ...   
4596                       1.142                           3   
4597                       1.555                           4   
4598                       1.404                      

In [4]:
df.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

In [5]:
# independent and dependent variables
X = df.iloc[: , :-1]
y = df.iloc[: , -1]

In [6]:
X

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0                0.00            0.00  ...                   0.0        0.000   
1                0.00            0.94  ...                   0.0        0.000   
2                0.64            0.25  ...                   0.0        0.010   
3                0.31            0.63  ...                   0.0        0.000   
4                0.31            0.63  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
4596             0.00            0.00  ...                   0.0        0.000   
4597             0.00            0.00  ...                   0.0        0.000   
4598             0.00            0.00  ...                   0.0        0.102   
4599             0.00            0.00  ...                   0.0        0.000   
4600             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0           0.000          0.0        0.778        0.000        0.000   
1           0.132          0.0        0.372        0.180        0.048   
2           0.143          0.0        0.276        0.184        0.010   
3           0.137          0.0        0.137        0.000        0.000   
4           0.135          0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...          ...   
4596        0.232          0.0        0.000        0.000        0.000   
4597        0.000          0.0        0.353        0.000        0.000   
4598        0.718          0.0        0.000        0.000        0.000   
4599        0.057          0.0        0.000        0.000        0.000   
4600        0.000          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537              

In [7]:
y

0       1
1       1
2       1
3       1
4       1
       ..
4596    0
4597    0
4598    0
4599    0
4600    0
Name: spamornot, Length: 4601, dtype: int64

In [8]:
y.value_counts()

0    2788
1    1813
Name: spamornot, dtype: int64

# spam (1) or not (0)

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

# Bernoulli classifier

In [10]:
classifier = BernoulliNB()
classifier.fit(X_train , y_train)

BernoulliNB()

In [11]:
y_pred = classifier.predict(X_test)

In [12]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.93      0.88      0.90       941
           1       0.82      0.89      0.85       578

    accuracy                           0.88      1519
   macro avg       0.87      0.88      0.88      1519
weighted avg       0.89      0.88      0.88      1519

[[824 117]
 [ 62 516]]
0.8821593153390388


# Multinomial classifier

In [13]:
classifier = MultinomialNB()
classifier.fit(X_train , y_train)

MultinomialNB()

In [14]:
y_pred = classifier.predict(X_test)

In [15]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.85      0.80      0.82       948
           1       0.69      0.77      0.73       571

    accuracy                           0.79      1519
   macro avg       0.77      0.78      0.78      1519
weighted avg       0.79      0.79      0.79      1519

[[754 194]
 [132 439]]
0.7853851217906518


# Gaussian classifier

In [16]:
classifier = GaussianNB()
classifier.fit(X_train , y_train)

GaussianNB()

In [17]:
y_pred = classifier.predict(X_test)

In [18]:
print(classification_report(y_pred , y_test))
print(confusion_matrix(y_pred ,y_test))
print(accuracy_score(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.73      0.95      0.83       681
           1       0.95      0.72      0.82       838

    accuracy                           0.82      1519
   macro avg       0.84      0.84      0.82      1519
weighted avg       0.85      0.82      0.82      1519

[[649  32]
 [237 601]]
0.8229098090849243


### in case of Bernoulli naive bayes classifier , we obtained an accuracy of 88.21 %

### in case of Multinomial naive bayes classifier , we obtained an accuracy of 78.53 %

### in case of Gaussian naive bayes classifier , we obtained an accuracy of 82.29 %

Thus, we get the best results in bernoulli classifier and we know that it works best in case of binary classification problems such as this problem where the two classes are the mail is a spam and it is not a spam